### PR TITLE
Add test for restore adjust and re-restore

### PR DESCRIPTION
**Pull request by KillianG**
add a test for reducing restore duration and a test for increasing restore duration

Issue: ZENKO-4727

### CODE REVIEW

File: tests/ctst/features/azureArchive.feature [hunk 1]
```diff
@@ -121,6 +121,38 @@ Feature: Azure Archive
...
+    Then object "obj-2" should be "cold" and have the storage class "e2e-azure-archive"
+    And object "obj1" should expire in 45 days
+    When i wait for 45 days
+    Then object "obj-1" should be "cold" and have the storage class "e2e-azure-archive"
+
+    Examples:
+        | versioningConfiguration | objectCount | objectSize | restoreDays |
+        |           Non versioned |           2 |        100 |           15 |
+        |               Versioned |           2 |        100 |           15 |
+        |               Suspended |           2 |        100 |           15 |
```
Comment (francoisferrand) [inline tests/ctst/features/azureArchive.feature:185]: as per the ticket, should also add a test case `restore→expiration→re-restore` (and verify it expires a second time as expected)

File: tests/ctst/common/common.ts [hunk 2]
```diff
@@ -258,18 +271,14 @@ Given('a transition workflow to {string} location', async function (this: Zenko,
...
-    const versionId = this.getSaved<Map<string, string>>('createdObjects')?.get(objName);
-    if (versionId) {
-        this.addCommandParameter({ versionId });
-    }
-    this.addCommandParameter({ restoreRequest: `Days=${days}` });
-    await S3.restoreObject(this.getCommandParameters());
+    await restoreObject.call(this, days, objectName);
 });
 
+When('i adjust restore duration of object {string} for {int} days',
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: "adjust restore" is not a separate step : this is not intrinsically an adjust, but only because of the condition of the object when this happens...

→ the steps should represent what we actually do (e.g. "issue an s3:restore command"), to make it easier to understand what goes on. I.e. we may have a "adjust restore" **_scenario_**, but that scenario would be simply "archive → restore → wait until restored → restore → check the restore duration has been updated"

File: tests/ctst/steps/azureArchive.ts [hunk 3]
```diff
@@ -373,7 +373,7 @@ When('i run sorbetctl to retry failed restore for {string} location', async func
 When('i wait for {int} days', { timeout: 10 * 60 * 1000 }, async function (this: Zenko, days: number) {
     const realTimeDay = days * 24 * 60 * 60 * 1000 /
         (this.parameters.TimeProgressionFactor > 1 ? this.parameters.TimeProgressionFactor : 1);
-    await Utils.sleep(realTimeDay + 200 * 1000); // We also wait for sorbet forwarder to see the object has expired
+    await Utils.sleep(realTimeDay); // We also wait for sorbet forwarder to see the object has expired
```
Comment (francoisferrand) [inline tests/ctst/steps/azureArchive.ts]: seems unrelated : if needed, should go in a separate commit to explain what happens here...

(and comment should probably be updated, since we don't have the "extra' delay?)

File: tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature [hunk 4]
```diff
@@ -125,52 +125,52 @@ Feature: Assume Role with Web Identity
...
+      | action                          | type                  | withVersioning | objectExists   | error                                           |
+      | RestoreObject                   | STORAGE_MANAGER       | with           | exists         | InvalidObjectState                              |
+      | GetBucketCors                   | STORAGE_MANAGER       | without        | does not exist | NoSuchCORSConfiguration                         |
+      | GetObjectLockConfiguration      | STORAGE_MANAGER       | without        | exists         | ObjectLockConfigurationNotFoundError            |
+      | GetObjectRetention              | STORAGE_MANAGER       | with           | exists         | Bucket is missing Object Lock Configuration     |
+      | GetObjectLegalHold              | STORAGE_MANAGER       | with           | exists         | Bucket is missing Object Lock Configuration     |
+      | PutObjectRetention              | STORAGE_MANAGER       | with           | exists         | MalformedXML: request xml is undefined or empty |
```
Comment (welansari) [inline tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature]: nit: for consistency with the other expected errors, should we only specify the error type `MalformedXML` without the description ?

File: tests/ctst/features/azureArchive.feature [hunk 5]
```diff
@@ -121,6 +121,70 @@ Feature: Azure Archive
         |               Versioned |           2 |      30000 |               1 |
         |               Suspended |           2 |      30000 |               1 |
 
+    @2.7.0
+    @PreMerge
+    @Flaky
+    @AzureArchive
+    @ColdStorage
+    @WIP
```
Comment (welansari) [inline tests/ctst/features/azureArchive.feature]: leftover from debugging ?

File: tests/ctst/features/azureArchive.feature [hunk 6]
```diff
@@ -121,6 +121,70 @@ Feature: Azure Archive
...
+        |           Non versioned |           2 |        100 |           15 |
+        |               Versioned |           2 |        100 |           15 |
+        |               Suspended |           2 |        100 |           15 |
+
+    @2.7.0
+    @PreMerge
+    @Flaky
+    @AzureArchive
+    @ColdStorage
+    @WIP
```
Comment (welansari) [inline tests/ctst/features/azureArchive.feature]: same here

**Review by welansari — APPROVED** [review]

File: solution/deps.yaml [hunk 7]
```diff
@@ -16,7 +16,7 @@ cloudserver:
   sourceRegistry: registry.scality.com/cloudserver
   dashboard: cloudserver-dashboards
   image: cloudserver
-  tag: 8.8.7
+  tag: 8.8.20
```
Comment (francoisferrand) [inline solution/deps.yaml]: already in #2023 I guess : will need to rebase after it is merged

**Review by francoisferrand — APPROVED** [review]

File: solution/deps.yaml [hunk 8]
```diff
@@ -99,7 +99,7 @@ vault:
 zenko-operator:
   sourceRegistry: registry.scality.com/zenko-operator
   image: zenko-operator
-  tag: 1.5.43
+  tag: 1.5.44
```
Comment (francoisferrand) [inline solution/deps.yaml]: not needed anymore, already bump'ed

File: tests/ctst/common/common.ts [hunk 9]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
         }
     });
 
-Then('kafka consumed messages should not take too much place on disk',
+Then('kafka consumed messages should not take too much place on disk', { timeout: 130000 },
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: timeout should should be based on `KafkaCleanerInterval`, i.e. a number of time the interval?

File: tests/ctst/common/common.ts [hunk 10]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
...
 
-        // Checking topics offsets before kafkacleaner passes to be sure kafkacleaner works
-        // This function can be improved by consuming messages and
-        // verify that the timestamp is not older than last kafkacleaner run
-        // Instead of waiting for a fixed amount of time,
-        // we could also check for metrics to see last kafkacleaner run
-        
-        // 10 seconds added to be sure kafkacleaner had time to process
-        await Utils.sleep(seconds * 1000 + 10000);
+        while (topics.length > 0 && new Date().getTime() - timeStart < 120000) {
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: what is the point of checking the duration?
* the timeout on the test is designed to catch test taking too long to complete
* here we simply want to try "a few times" (or even not limit, and just fail without timeout if it never settles...)

File: tests/ctst/common/common.ts [hunk 11]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
...
+            const previousOffsets = await getTopicsOffsets(topics, kafkaAdmin);
 
-        const newOffsets = await getTopicsOffsets(topics, kafkaAdmin);
+            const seconds = parseInt(this.parameters.KafkaCleanerInterval);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: no point doing this in the loop, just compute once before the loop

File: tests/ctst/common/common.ts [hunk 12]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
...
+                        `Topic ${topics[i]} partition ${j} offset has not increased,
+                        previousOffsets: ${previousOffsets[i].partitions[j].low} /\
+                         ${previousOffsets[i].partitions[j].high},
+                        newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+                    topics.splice(i, 1);
+                }
             }
         }
+
+        assert(topics.length === 0, `Topics ${topics.join(', ')} still have not been cleaned`);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: here we should fail if we exited the loop with the "`newMessagesAfterClean`" case, i.e. if it never converged

File: tests/ctst/common/common.ts [hunk 13]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
...
+            await Utils.sleep(seconds * 1000 + 10000);
+
+            const newOffsets = await getTopicsOffsets(topics, kafkaAdmin);
+
+            for (let i = 0; i < topics.length; i++) {
+                process.stdout.write(`\nChecking topic ${topics[i]}\n`);
+                for (let j = 0; j < newOffsets[i].partitions.length; j++) {
+                    // Checking that the min offset has increased due to kafkacleaner
+                    // or that it didn't need to change because there was no new messages
+                    // or new messages came after a clean but still cleaned to previous high offset
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: the comment is hard to relate with the code now...

File: tests/ctst/common/common.ts [hunk 14]
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
...
+                    const lowOffsetIncreased = newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low;
+                    const noNewMessages = newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low;
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: not used here → move after the `if` for clarify, where they will be used

File: tests/ctst/common/common.ts (hunk 10 above)
Comment (KillianG) [reply tests/ctst/common/common.ts]: This is to end properly and not getting killed by the timeout so at the end we can log each topic that is not cleaned in time, if we just wait for the timeout, we'll end up with no readable logs.. just "test timed out" with no info..

File: tests/ctst/common/common.ts (hunk 9 above)
Comment (KillianG) [reply tests/ctst/common/common.ts]: Should be great but unfortuneatly we can't access `Zenko` object here :(
And I am not sure if this is great anyway, because if we set kafkacleaner interval to 10mn for some reason, we'll end up with 30mn timeout ?

File: tests/ctst/common/common.ts (hunk 12 above)
Comment (KillianG) [reply tests/ctst/common/common.ts]: Maybe i missed something, but this is already what it does, while there is `newMessagesAfterClean` we continue (i.e we don't remove the topic from the list of topics to be checked) and here, if there is topics that remains in the array, it's that they haven't been cleaned and we print them to be able to debug in case it happens

File: tests/ctst/common/common.ts [hunk 15]
```diff
@@ -300,40 +300,68 @@ Then('i {string} be able to add user metadata to object {string}',
...
-            for (let j = 0; j < newOffsets[i].partitions.length; j++) {
-                // Checking that the min offset has increased due to kafkacleaner
-                // or that it didn't need to change because there was no new messages
-                assert.ok(newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low ||
-                    newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low,
-                `Topic ${topics[i]} partition ${j} offset has not increased,
-                previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
-                newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+
+
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: nit: extra empty line

File: tests/ctst/common/common.ts [hunk 16]
```diff
@@ -300,40 +300,68 @@ Then('i {string} be able to add user metadata to object {string}',
...
+                        previousOffsets: ${previousOffsets[i].partitions[j].low} /\
+                         ${previousOffsets[i].partitions[j].high},
+                        newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+
+                    // Topic is cleaned, we don't need to check it anymore
+                    topics.splice(i, 1);
+                }
             }
         }
+        clearTimeout(timeoutID);
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts]: in case of `assert` before, this will not be executed : which could cause strange behavior.

a safer approach would be to use `try{}finally{}` block:

```suggestion
        const timeoutID = setTimeout(....);
        try {
           //run the check
        } finally {
           clearTimeout(timeoutID);
        }
```

(atlernatively, we could maybe [`async.timeout`](https://caolan.github.io/async/v3/docs.html#timeout) instead)

File: tests/ctst/common/common.ts [hunk 17]
```diff
@@ -300,40 +300,72 @@ Then('i {string} be able to add user metadata to object {string}',
...
-            for (let j = 0; j < newOffsets[i].partitions.length; j++) {
-                // Checking that the min offset has increased due to kafkacleaner
-                // or that it didn't need to change because there was no new messages
-                assert.ok(newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low ||
-                    newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low,
-                `Topic ${topics[i]} partition ${j} offset has not increased,
-                previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
-                newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+
+        try {
```
Comment (francoisferrand) [inline tests/ctst/common/common.ts:311]: should encompass listing of kafka topic as well, i.e. start immediately after creating the timer

**Review by francoisferrand — APPROVED** [review]
